README.md
# changelogger

Changelogger is a CLI tool for maintaining changelog files and version releases.

## Installation

To install it, you can either use the following one-liner or download the [binary](https://github.com/MarkusFreitag/changelogger/releases/latest/download/changelogger_linux_amd64.deb), [deb](https://github.com/MarkusFreitag/changelogger/releases/latest/download/changelogger_linux_amd64.rpm) or [rpm](https://github.com/MarkusFreitag/changelogger/releases/latest/download/changelogger_linux_amd64.tar.gz) package from the releases.
```bash
curl -Ls https://git.io/changelogger.sh | sh
```

When you using it, every 24h, it checks whether new updates are available. If so, it can be updated using its `update` command.
```bash
changelogger update
```

## Usage

### Add a new entry

To initialize the file or create a new entry, simply run the tool without additional commands. It will open an editor where you can enter your changes. In the editor you will see the recent changes for your user. Just add leave these lines and add your new ones below it.

### Make a new release

You will be asked, what kind of version bump you would like to do. Then the last version will be bumped and set for the last unreleased entry. This will not only update the version but also the author and the timestamp.
```bash
changelogger release new
```


CHANGELOG.md
# Patch Release v0.6.1 (2023-11-23)
  * **Tom Siewert**
    * debian: lowercase unstable

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Minor Release v0.6.0 (2023-01-18)
  * **Markus Freitag**
    * pkg/gitconfig: replace character-based with ini config parser

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.5.3 (2022-08-23)
  * **Tom Siewert**
    * pkg/gitconfig: Add ~/.config/git/config to possible files
    * ci: Use official golangci-lint action

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.5.2 (2022-05-10)
  * **Tom Siewert**
    * ci: Migrate to GitHub Workflows
    * install.sh: Update to latest to support arm64 properly

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.5.1 (2022-03-23)
  * **Markus Freitag**
    * add arm64 builds
    * update module dependencies

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Minor Release v0.5.0 (2020-10-19)
  * **Markus Freitag**
    * - cmd/debian: Add version and date flag for experimental releases

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.4.3 (2020-04-18)
  * **Markus Freitag**
    * Move darwin from goarch to goos

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.4.2 (2020-04-18)
  * **Markus Freitag**
    * Add binary builds for darwin

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.4.1 (2020-03-26)
  * **Markus Freitag**
    * Add missing whitespaces to debian changelog template

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Minor Release v0.4.0 (2020-03-05)
  * **Markus Freitag**
    * Drop auto update checker
    * Prefix changelog entry with an asterisk if missing
    * Remove empty trailing lines
    * Unify error handling

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.3.3 (2020-02-12)
  * **Markus Freitag**
    * Fix goreleaser config

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.3.2 (2020-02-12)
  * **Markus Freitag**
    * **bugfix** Fix early return when generating debian changelog

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.3.1 (2019-07-30)
  * **Markus Freitag**
    * Remove 'v' from version string when generating debian changelog

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Minor Release v0.3.0 (2019-07-26)
  * **Markus Freitag**
    * Add CLI command `debian`
      * `debian dummy` creates a changelog file containing only the latest
        release with the hint to check CHANGELOG.md
      * `debian full` generates a debian formated changelog out of CHANGELOG.md

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.2.5 (2019-07-24)
  * **Markus Freitag**
    * Sort authors within a release

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.2.4 (2019-05-06)
  * **Markus Freitag**
    * Set default versioning to semver

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.2.3 (2019-02-12)
  * **Markus Freitag**
    * Enable syntax highlighting in editor
    * Run update check just every 24h

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.2.2 (2019-02-01)
  * **Markus Freitag**
    * Do not exit program when update check fails

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Patch Release v0.2.1 (2019-02-01)
  * **Markus Freitag**
    * Bugfix for the update checking routine

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Minor Release v0.2.0 (2019-02-01)
  * **Markus Freitag**
    * Implement update routine
      * automatic update checker in the init phase
      * `update` command to selfupdate the binary
    * Implement `json` command
    * Add prerun check for minimal version

*Released by Markus Freitag <fmarkus@mailbox.org>*

# Minor Release v0.1.0 (2019-01-31)
  * **Markus Freitag**
    * Initial release, implemented features
      * create or update CHANGELOG files
      * bump version for a new release
      * show information about the last release

*Released by Markus Freitag <fmarkus@mailbox.org>*


cmd/debian.go
// Copyright © 2019 Markus Freitag <fmarkus@mailbox.org>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package cmd

import (
	"bytes"
	"errors"
	"fmt"
	"io/ioutil"
	"os"
	"strings"
	"text/template"

	"github.com/MarkusFreitag/changelogger/pkg/parser"
	"github.com/MarkusFreitag/changelogger/pkg/stringutil"
	"github.com/spf13/cobra"
)

var (
	changelogPath   string
	templatePath    string
	binaryName      string
	customVersion   string
	customDate      string
	temp            *template.Template
	releases        parser.Releases
	out             = bytes.NewBufferString("")
	defaultTemplate = `{{.BinaryName}} ({{.Version}}) unstable; urgency=medium

  {{.Text}}

 -- {{.AuthorName}} <{{.AuthorMail}}>  {{.Date}}

`
)

func loadTemplate() (*template.Template, error) {
	if _, err := os.Stat(templatePath); err == nil {
		customTemplate, err := ioutil.ReadFile(templatePath)
		if err != nil {
			return nil, err
		}
		return template.New("changelog").Parse(string(customTemplate))
	}
	return template.New("changelog").Parse(defaultTemplate)
}

type templateData struct {
	BinaryName string
	Version    string
	Text       string
	AuthorName string
	AuthorMail string
	Date       string
}

var debianCmd = &cobra.Command{
	Use:   "debian",
	Short: "Generate debian changelog file",
	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
		if _, err := os.Stat(changelogPath); !os.IsNotExist(err) && !force {
			return fmt.Errorf("`%s` already exist, use --force to overwrite it", changelogPath)
		}
		if value, ok := os.LookupEnv("DEBIAN_NAME"); ok {
			binaryName = value
		} else {
			return errors.New("`DEBIAN_NAME` env var is not set")
		}

		var err error
		temp, err = loadTemplate()
		if err != nil {
			return err
		}

		releases, err = parser.ReadFile(changelogFile)
		return err
	},
	PersistentPostRunE: func(cmd *cobra.Command, args []string) error {
		if _, err := os.Stat(templatePath); err == nil {
			err := os.Remove(templatePath)
			if err != nil {
				return err
			}
		}
		return ioutil.WriteFile(changelogPath, out.Bytes(), 0644)
	},
}

var debianDummyCmd = &cobra.Command{
	Use:   "dummy",
	Short: "Generate dummy debian changelog",
	RunE: func(cmd *cobra.Command, args []string) error {
		var rel *parser.Release
		for _, r := range releases {
			if r.Released {
				rel = r
				break
			}
		}
		if rel == nil {
			return errors.New("no released version found")
		}

		data := templateData{
			BinaryName: binaryName,
			AuthorName: rel.By.Name,
			AuthorMail: rel.By.Email,
			Text:       "* See CHANGELOG.md",
		}
		if customVersion != "" {
			data.Version = customVersion
		} else {
			data.Version = strings.TrimPrefix(rel.Version.Original(), "v")
		}
		if customDate != "" {
			data.Date = customDate
		} else {
			data.Date = rel.Date.Format("Mon, 02 Jan 2006 15:04:05 -0700")
		}

		return temp.ExecuteTemplate(out, "changelog", data)
	},
}

var debianFullCmd = &cobra.Command{
	Use:   "full",
	Short: "Generate full debian changelog",
	RunE: func(cmd *cobra.Command, args []string) error {
		for _, rel := range releases {
			if !rel.Released {
				continue
			}

			var text string
			if value, ok := rel.Changes[rel.By.Name]; ok && len(rel.Changes) == 1 {
				text = value
			} else {
				for _, author := range rel.Changes.SortedAuthors() {
					text += fmt.Sprintf("[ %s ]\n", author)
					text += stringutil.IncrIndent(rel.Changes[author], 2)
					text += "\n"
				}
			}
			text = strings.TrimSuffix(text, "\n")
			text = stringutil.IncrIndent(text, 2)

			data := templateData{
				BinaryName: binaryName,
				AuthorName: rel.By.Name,
				AuthorMail: rel.By.Email,
				Text:       text,
			}
			if customVersion != "" {
				data.Version = customVersion
			} else {
				data.Version = strings.TrimPrefix(rel.Version.Original(), "v")
			}
			if customDate != "" {
				data.Date = customDate
			} else {
				data.Date = rel.Date.Format("Mon, 02 Jan 2006 15:04:05 -0700")
			}

			err := temp.ExecuteTemplate(out, "changelog", data)
			if err != nil {
				return err
			}

			out.WriteString("\n")
		}
		return nil
	},
}

func init() {
	rootCmd.AddCommand(debianCmd)
	debianCmd.PersistentFlags().StringVar(&changelogPath, "output-file", "debian/changelog", "debian changelog path")
	debianCmd.PersistentFlags().StringVar(&templatePath, "template-file", "debian/changelog.template", "template path")
	debianCmd.PersistentFlags().BoolVar(&force, "force", false, "overwrite an existing file")
	debianCmd.PersistentFlags().StringVar(&customVersion, "version", "", "specify custom version for experimental releases")
	debianCmd.PersistentFlags().StringVar(&customDate, "date", "", "specify custom data for experimental releases")

	debianCmd.AddCommand(debianDummyCmd)
	debianCmd.AddCommand(debianFullCmd)
}
